### ISSUE TITLE
Synchronization seems to go into infinite loop

### ISSUE THREAD

**Issue by lucatorella** [couchbase/couchbase-lite-ios#1273]
I just tried out the newer couch base version (1.2.1) and it keeps synchronizing forever. It seems it goes into recursive infinite loop.
I'm able to reproduce the same behavior on the GrocerySync app just by adding our credentials.

This is part of the app log:
```
...
2016-06-06 14:45:28.513 GrocerySync[1309:4928114] SYNC progress: 8167 / 10271
2016-06-06 14:45:28.784 GrocerySync[1309:4928114] SYNC progress: 8277 / 10271
2016-06-06 14:45:29.060 GrocerySync[1309:4928114] SYNC progress: 8400 / 10271
2016-06-06 14:45:29.337 GrocerySync[1309:4928114] SYNC progress: 8518 / 10271
2016-06-06 14:45:29.610 GrocerySync[1309:4928114] SYNC progress: 8633 / 10271
2016-06-06 14:45:29.882 GrocerySync[1309:4928114] SYNC progress: 8741 / 10271
2016-06-06 14:45:30.160 GrocerySync[1309:4928114] SYNC progress: 8863 / 10271
2016-06-06 14:45:30.431 GrocerySync[1309:4928114] SYNC progress: 8984 / 10271
2016-06-06 14:45:30.709 GrocerySync[1309:4928114] SYNC progress: 9102 / 10271
2016-06-06 14:45:30.983 GrocerySync[1309:4928114] SYNC progress: 9225 / 10271
2016-06-06 14:45:31.256 GrocerySync[1309:4928114] SYNC progress: 9338 / 10271
2016-06-06 14:45:31.521 GrocerySync[1309:4928114] SYNC progress: 9403 / 10271
2016-06-06 14:45:31.807 GrocerySync[1309:4928114] SYNC progress: 9404 / 10271
2016-06-06 14:45:32.072 GrocerySync[1309:4928114] SYNC progress: 9472 / 10271
2016-06-06 14:45:32.338 GrocerySync[1309:4928114] SYNC progress: 9535 / 10271
2016-06-06 14:45:32.610 GrocerySync[1309:4928114] SYNC progress: 9645 / 10271
2016-06-06 14:45:32.882 GrocerySync[1309:4928114] SYNC progress: 9752 / 10271
2016-06-06 14:45:33.187 GrocerySync[1309:4928114] SYNC progress: 9864 / 10271
2016-06-06 14:45:33.458 GrocerySync[1309:4928114] SYNC progress: 9968 / 10271
2016-06-06 14:45:34.526 GrocerySync[1309:4928114] SYNC progress: 19254 / 19656
2016-06-06 14:45:35.428 GrocerySync[1309:4928114] SYNC progress: 28907 / 29484
2016-06-06 14:45:35.447 GrocerySync[1309:4928114] SYNC progress: 28960 / 29484
2016-06-06 14:45:35.720 GrocerySync[1309:4928114] SYNC progress: 29076 / 29484
2016-06-06 14:45:35.929 GrocerySync[1309:4928114] SYNC progress: 29086 / 29484
2016-06-06 14:45:35.984 GrocerySync[1309:4928114] SYNC progress: 29202 / 29484
2016-06-06 14:45:37.125 GrocerySync[1309:4928114] SYNC progress: 39123 / 39312
2016-06-06 14:45:38.009 GrocerySync[1309:4928114] SYNC progress: 48923 / 49140
2016-06-06 14:45:38.022 GrocerySync[1309:4928114] SYNC progress: 48968 / 49140
2016-06-06 14:45:38.927 GrocerySync[1309:4928114] SYNC progress: 58768 / 58968
2016-06-06 14:45:38.945 GrocerySync[1309:4928114] SYNC progress: 58821 / 58968
2016-06-06 14:45:39.842 GrocerySync[1309:4928114] SYNC progress: 68721 / 68796
...
```

If I enable SyncVerbose logging, I see that it's going over and over the same documents, as you can see when I do the following grep:
```
cat log.txt | grep "1c194c15dd49ed4b283f728e149db7fc"
15:07:36.393‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:07:48.305‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:07:59.152‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:08:10.775‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:08:22.420‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:08:42.785‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:08:56.727‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:09:10.133‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
15:09:22.184‖ SyncVerbose: CBLRestPuller[https://xxx-data1.xxx.com/customer_3006/]: Received #14068 {1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}
    "{1c194c15dd49ed4b283f728e149db7fc #2-3f91ead3e43e7c6eed0a0df00c225edd}",
```

This is the server log:
```
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:08:40 +0200]  "PUT /customer_3006/_local/567923f8c7d0934cbeb0a020f3e55f06f61781f7 HTTP/1.1" 201 80 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 0.002 0.002 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:08:47 +0200]  "POST /customer_3006/_changes HTTP/1.1" 200 285618 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 6.970 0.165 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:08:54 +0200]  "PUT /customer_3006/_local/567923f8c7d0934cbeb0a020f3e55f06f61781f7 HTTP/1.1" 201 80 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 0.002 0.002 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:08:59 +0200]  "POST /customer_3006/_changes HTTP/1.1" 200 285618 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 5.286 0.164 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:09:07 +0200]  "PUT /customer_3006/_local/567923f8c7d0934cbeb0a020f3e55f06f61781f7 HTTP/1.1" 201 80 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 0.002 0.002 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:09:15 +0200]  "POST /customer_3006/_changes HTTP/1.1" 200 285618 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 7.268 0.166 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:09:20 +0200]  "PUT /customer_3006/_local/567923f8c7d0934cbeb0a020f3e55f06f61781f7 HTTP/1.1" 201 80 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 0.003 0.003 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:09:25 +0200]  "POST /customer_3006/_changes HTTP/1.1" 200 285626 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 5.357 0.170 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:10:56 +0200]  "PUT /customer_3006/_local/567923f8c7d0934cbeb0a020f3e55f06f61781f7 HTTP/1.1" 201 80 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 0.003 0.003 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:11:04 +0200]  "POST /customer_3006/_changes HTTP/1.1" 200 285618 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 7.928 0.167 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:11:19 +0200]  "PUT /customer_3006/_local/567923f8c7d0934cbeb0a020f3e55f06f61781f7 HTTP/1.1" 201 80 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 0.003 0.003 .
xxx.xx.xx.xxx - device_8496 [06/Jun/2016:15:11:25 +0200]  "POST /customer_3006/_changes HTTP/1.1" 200 285618 "-" "CouchbaseLite/1.2 (iOS 1.2.1)" 6.027 0.200 .
```
--------
* Version:   1.2.1
* Client OS: iOS 9.3.2
* Server:    couchdb 1.6.1

**Comment by lucatorella** [comment]
Here you can find part of the log: https://gist.github.com/lucatorella/68a823be9e454e696ff0b3bf76f730bc

**Comment by snej** [comment]
This is a known compatibility problem with CouchDB; it's been fixed on the `release/1.2.2` branch. I don't know if we're going to have an official 1.2.2 release before 1.3, but you can check out and build that branch yourself. Sorry for the inconvenience.
